### PR TITLE
Refactor shell code to Go

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Description
Transform bash script into Go. The best things seems to be performance improvement, maintainability/readability and args and flags validation.

## How to test
Run the script with `go run docs/examples/getDataMoverResources.go [flags]` and compare with the bash one.

### CODE REVIEW

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:57]: delete this func

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:97]: I wanted to add validations, but the way the script is intended to run I think it is not good. With `watch`, it would run this every iteration. Better to remove, right?

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:204]: backup and restore should be required, but there is not this feature in cobra yet (but it is already merged)

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:1]: I think my biggest question is how top run the script. I saw this discussion https://gist.github.com/posener/73ffd326d88483df6b1cb66e8ed1e0bd

what you suggest?

**Comment by mateusoliveira43** [comment]
@weshayutin can you check if I am going in the right direction? `-b` should be already working, the others ones should be simple to implement, just want to check first

**Review by mpryc — COMMENTED** [review]
@mateusoliveira43 my general question around this script is if you have considered using `go` libraries to access the k8s / OCP cluster instead of running shell that invokes `oc` command?

Here are few examples:
  https://github.com/kubernetes/client-go/tree/master/examples

**Comment by mateusoliveira43** [comment]
@mpryc did not thought about that. I will take a look and test out

File: docs/examples/getDataMoverResources.go
Comment (kaovilai) [reply docs/examples/getDataMoverResources.go:1]:
```
❯ go run docs/examples/getDataMoverResources.go --help
Check the OADP DataMover CR's in real time.

You need to log in to your cluster and deploy OADP prior to run this script

Usage:
  getDataMoverResources [flags]

Examples:
go run docs/examples/getDataMoverResources.go -b

Flags:
  -b, --backup    Check a backup
  -d, --details   Print a list of the relevant backup/restore CR's
  -h, --help      help for getDataMoverResources
  -r, --restore   Check a restore

```

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:90]: tested @mpryc suggestion. I liked :smile:  what you prefer so I can apply to the remaining: shell or Kubernetes lib?

Comment (kaovilai) [reply docs/examples/getDataMoverResources.go:90]: go lib would be more reliable but adds go dependency that could come back to haunt us.

Don't go depend on something not meant for consumption like [k8s.io/kubectl/pkg/cmd/cp](https://gitlab.cee.redhat.com/app-mig/oadp-e2e-qe/-/commit/1281a66f866bf39b2a85b450302af5f7c7e1c374#a1f4f20d8f21da3825a49a53c8146f0230a76a67_39_36) tho :P

Comment (mateusoliveira43) [reply docs/examples/getDataMoverResources.go:90]: yeah, that is a valid concern. Even without Kubernetes lib, we are using cobra, which is not direct as well

since Go does not have development dependencies, I think the best solution would be a separate repo and use [Makefile go-install-tool](https://github.com/openshift/oadp-operator/blob/master/Makefile#L284-L296) (which can become 2 repos changes to achieve a simple change in the future). Do you suggest another solution to overcome Go dependency problem?

Comment (kaovilai) [reply docs/examples/getDataMoverResources.go:90]: I mean you could still consume kubectl via go like in https://gitlab.cee.redhat.com/app-mig/oadp-e2e-qe/-/blob/master/utils/pods/kube_exec.go#L37

Just consume the top main cobra (kubectl) command, rather than digging into subcommand for example.

I'm not as worried about client-go lib as we import that already.

Comment (mateusoliveira43) [reply docs/examples/getDataMoverResources.go:90]: I think I still did not understand your suggestion @kaovilai 

do you suggest to have a function like this?
```go
func kubectl(args ...string) string {
  kubectl := cmd.NewDefaultKubectlCommand()
  kubectl.SetArgs(args...)
  outputBuffer := new(bytes.Buffer)
  kubectl.SetOut(outputBuffer)
  err := kubectl.Execute()
  exitOnError(err)
  return outputBuffer.String()
}

kubectl([]string{"get", "volumesnapshot", "-A"})
```

Comment (kaovilai) [reply docs/examples/getDataMoverResources.go:90]: Correct.. if you are to do something with kubectl, that would be easier/faster to consume and reliable (shell independent).

Comment (kaovilai) [reply docs/examples/getDataMoverResources.go:90]: if you can use client-go instead, that would be preferable but otherwise it's fine.
https://stackoverflow.com/questions/43314689/example-of-exec-in-k8ss-pod-by-using-go-client

Comment (kaovilai) [reply docs/examples/getDataMoverResources.go:90]: oc get volumesnapshots -A should definitely use client-go, you can call length on the result easily.

Comment (mateusoliveira43) [reply docs/examples/getDataMoverResources.go:90]: alright, thanks! going to study it and make the required changes

Comment (mateusoliveira43) [reply docs/examples/getDataMoverResources.go]: used go client for everything :) resolving conversation, but feel free to open it again if you see any problems

File: docs/examples/datamover_resources.sh
Comment (mateusoliveira43) [inline docs/examples/datamover_resources.sh:1]: TODO run grep file_name and see if any place needs changes

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:141]: typo then -> then

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:204]: is there a less verbose way of getting properties from object?

File: docs/examples/datamover_resources.sh
Comment (mateusoliveira43) [reply docs/examples/datamover_resources.sh:1]: found a problem: how to replace this https://github.com/openshift/oadp-operator/blob/master/docs/oadp_cheat_sheet.md?plain=1#L230 ?

is there something shorter than this?
```sh
git clone --depth=1 git@github.com:openshift/oadp-operator.git
go build -C oadp-operator docs/examples/getDataMoverResources.go
mv oadp-operator/getDataMoverResources ~/.local/bin/getDataMoverResources
rm -r oadp-operator
```

do I need to change anything in PROW? I thing I remember a regex for this file

File: docs/examples/getDataMoverResources.go
Comment (kaovilai) [reply docs/examples/getDataMoverResources.go:204]: what do you mean less verbose?

Comment (mateusoliveira43) [reply docs/examples/getDataMoverResources.go:204]: something like `volumeSnapshotBackup.Object["status"]["phase"]` instead of `volumeSnapshotBackup.Object["status"].(map[string]interface{})["phase"]`

File: docs/examples/datamover_resources.sh
Comment (mateusoliveira43) [reply docs/examples/datamover_resources.sh:1]: wait this PR https://github.com/openshift/oadp-operator/pull/1103 before merging this one?

**Comment by weshayutin** [comment]
@mateusoliveira43 let's wrap this up vs. letting it sit if you have some time.  Ready for testing?

**Comment by mateusoliveira43** [comment]
Yeah, already ready @weshayutin 
I will just fix conflict during the day

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:73]: will test in my upcoming tasks

Comment (mateusoliveira43) [reply docs/examples/getDataMoverResources.go:73]: tested

it works (it sees volumeSnapshotContents) but could be better

should I 
- merge like this (focused in OLD DataMover)
- add NEW DataMover stuff (DataUpload and DataDownload) and add new flag to select which to use
- remove OLD DataMover stuff (VSM and Volsync), and add NEW DataMover stuff
?

**Comment by shubham-pampattiwar** [comment]
@mateusoliveira43 Currently this PR carters to OADP 1.2 datamover (Volsync + VSM based) , as we are moving away from this datamover to support Velero built-in datamover in OADP 1.3, wouldn't it make sense to make this work with the velero built-in datamover ?

**Comment by weshayutin** [comment]
@mateusoliveira43 might as well create a flag for 1.2 vs. 1.3.  In 1.3. track DUCR's and DDCR's.

File: docs/examples/getDataMoverResources.go
Comment (mateusoliveira43) [inline docs/examples/getDataMoverResources.go:130]: Added flag to select between Velero Built-in DataMover and VSM/Volsync DataMover :smile:

File: docs/examples/debug.md
Comment (kaovilai) [inline docs/examples/debug.md:5]: btw `go run` is just a quick and dirty way to run from go code directly.

For a published code, it would be good to `go build getDataMoverResources.go -o getDataMoverResources` then use compiled binary for faster speed.
```
./getDataMoverResources
```

Comment (mateusoliveira43) [reply docs/examples/debug.md:5]: oh yeah, forget that, I'm used to Python ehhehe Going to add in the doc to user to compile first (and also tell about flag to switch between 1.2 and 1.3 datamover)

**Comment by mrnold** [comment]
Who is the intended audience for this section of docs? This change is fine if it's meant for OADP developers, but are we also presenting these examples to users or product support personnel anywhere? I don't think the second group will want to have to compile a go program instead of copying a shell script.

**Review by mrnold — APPROVED** [review]

**Comment by mateusoliveira43** [comment]
/hold

We will wait to have a easy way to users to run the script (without the need of having Go in the machines)

**Comment by kaovilai** [comment]
> We will wait to have a easy way to users to run the script (without the need of having Go in the machines)

We can prebuilt the binaries for download.

**Comment by mateusoliveira43** [comment]
Yeah, I think we also brainstormed container images, @kaovilai 

But will wait a future PR that Michal is working on, to see if there is an easier way

**Comment by mateusoliveira43** [comment]
gave up on this

**Pull request closed without merging**
